Tabs: indicator animation

DaniGuardiola opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Makes the `Tabs` indicator animated (except when the user prefers reduced motion).

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Looks cool. @mirka can probably add more context :)

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

See styles. Position and length is synced through the Ariakit state store.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

Play with it on Storybook. Optionally simulate the "prefers reduced motion" setting through devtools.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

Tab into the component, then use arrow keys to change tabs.

## Screenshots or screencast <!-- if applicable -->

https://github.com/WordPress/gutenberg/assets/8862967/c03f6abc-f331-4711-b94d-8c801b41deaf

In Windows high-contrast mode:

https://github.com/WordPress/gutenberg/assets/8862967/9c11703a-7fda-4a4d-8445-e704543d4457

DaniGuardiola committed:
Initial tab indicator animation implementation

DaniGuardiola committed:
Add changelog entry

DaniGuardiola committed:
Minor tweak

DaniGuardiola committed:
Merge branch 'trunk' into tabs-animation

mirka commented on packages/components/src/tabs/styles.ts:
```diff
@@ -13,12 +13,31 @@ import { space } from '../utils/space';
+			${ COLORS.theme.accent };
+		pointer-events: none;
+		@media not ( prefers-reduced-motion: reduce ) {
+			transition: left 0.2s ease-out;
```
Let's add a `width` transition too, it becomes quite noticeable when the tab widths are different.

DaniGuardiola replied:
Good catch.

mirka commented on packages/components/src/tabs/styles.ts:
```diff
@@ -13,12 +13,31 @@ import { space } from '../utils/space';
+		border-bottom: var( --wp-admin-border-width-focus ) solid
+			${ COLORS.theme.accent };
+		pointer-events: none;
+		@media not ( prefers-reduced-motion: reduce ) {
```
I actually prefer this "**do if doesn't prefer** reduced motion" model, but there are "**undo if prefers** reduced motion" mixins like [`reduceMotion`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/utils/reduce-motion.js) and its [Sass counterpart](https://github.com/WordPress/gutenberg/blob/af2226259f60b019e7a8673a08d573db08f9ade3/packages/base-styles/_mixins.scss#L185) that are widely used in the package.

I see the argument for sticking to a single mental model for dealing with reduced motion throughout the codebase, but I don't necessarily love the approach that the mixins take. I feel like it's the opposite of how CSS should be architected.

Curious to hear your thoughts on this. Any pros/cons I'm missing? I'm ok with phasing out the mixin usage, at least within the components package, if that makes more sense.

DaniGuardiola replied:
I saw those, but I don't really like them personally. They are abstractions that save a couple of words worth of keystrokes and obscure the actual output, plus they add to the bundle size. I find them a bit confusing, and as with most abstractions, there's not a lot of margin for nuance. Reduced motion doesn't mean no motion, and in some cases it makes sense to keep some transitions even if the user has enabled this setting.

I just don't see a reason to abstract `@media not ( prefers-reduced-motion: reduce )` (45 characters, spaces and all) in the first place :P

I do appreciate consistency, but I'd definitely push for phasing these out. Is that something we could do?

mirka replied:
Yes, let's do that! Please review this PR, especially that the suggested alternatives makes sense: #60839

DaniGuardiola replied:
Awesome!

mirka commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -17,19 +17,36 @@ import type { TabListProps } from './types';
+	useEffect( () => {
+		if ( selectedTabEl )
+			setIndicatorPosition( {
+				left: selectedTabEl.offsetLeft,
```
This also needs to work in `orientation="vertical"`.

(Not that I think the indicator styling for vertical tabs is good right now — we'll probably want to improve it at some point — but for the scope of this PR let's just make sure it isn't broken.)

DaniGuardiola replied:
Oh, I didn't see any story for this so I assumed it wasn't officially supported. I'll take a look.

DaniGuardiola replied:
Added.

mirka commented on packages/components/src/tabs/styles.ts:
```diff
@@ -13,12 +13,31 @@ import { space } from '../utils/space';
+			${ COLORS.theme.accent };
+		pointer-events: none;
+		@media not ( prefers-reduced-motion: reduce ) {
+			transition: left 0.2s ease-out;
```
It think we can make the focus-visible ring transition a bit more congruent with the new transitions. It feels too abrupt.

https://github.com/WordPress/gutenberg/blob/ecf24f9b8adc4147ef0b50e7bcfec3f4187f7351/packages/components/src/tabs/styles.ts#L92-L103

☝️ Switching this to a transition between opacity 0 and 1 feels better to me, rather than the current `transition: all` that transitions the entire box-shadow in a clunky way. What do you think?

DaniGuardiola replied:
Done. Feels much better indeed 👍

mirka commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -17,19 +17,36 @@ import type { TabListProps } from './types';
+				left: selectedTabEl.offsetLeft,
+				width: selectedTabEl.getBoundingClientRect().width,
+			} );
+	}, [ selectedTabEl ] );
```
This is not going to be enough because the selected tab width can change dynamically.

You can see this kind of behavior in the block editor, for example:

https://github.com/WordPress/gutenberg/assets/555336/de3e3eec-12ac-45fb-8a22-7b0aada2c36f

We have a [`ResizeObserver` utility hook](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-compose/#useresizeobserver) that could be used.

DaniGuardiola replied:
Implemented using ResizeObserver (not the linked hook because their API does not work here).

mirka commented on packages/components/src/tabs/styles.ts:
```diff
@@ -13,12 +13,31 @@ import { space } from '../utils/space';
 import { reduceMotion } from '../utils/reduce-motion';
 
 export const TabListWrapper = styled.div`
+	position: relative;
```
`isolation: isolate` is probably better here. I already found a bad interaction with a downstream override in the Preferences modal 🙁

https://github.com/WordPress/gutenberg/blob/f003f28c1d1b01d45a0b3455bd90cb671b7a993f/packages/preferences/src/components/preferences-modal-tabs/style.scss#L3-L4

DaniGuardiola replied:
Hmm this should be fine though, the only override that'd break it is `position: static`, am I missing something?

DaniGuardiola replied:
TIL, but `isolation: isolate` does not work. Seems like the stacking context is unrelated to positioning. According to MDN positioning is relative to the root OR the closest non-statically-positioned ancestor. So we need to stick with `position: relative`.

mirka commented on packages/components/src/tabs/styles.ts:
```diff
@@ -13,12 +13,31 @@ import { space } from '../utils/space';
 	&[aria-orientation='vertical'] {
 		flex-direction: column;
 	}
+	&::after {
```
This is unfortunately one of those times where we need to deal with downstream overrides just because we live in the same repo.

The Preferences modal can be shown from the dropdown menu in the top right of the block editor.

https://github.com/WordPress/gutenberg/blob/f003f28c1d1b01d45a0b3455bd90cb671b7a993f/packages/preferences/src/components/preferences-modal-tabs/style.scss#L32-L34

## Expected

(Don't mind the different accent color, that's just due to the theme.)

<img width="787" alt="Preferences panel in trunk" src="https://github.com/WordPress/gutenberg/assets/555336/ed0d9a36-fafd-46fa-b151-1ff49098d96d">

## Actual

<img width="777" alt="Preferences modal in this PR" src="https://github.com/WordPress/gutenberg/assets/555336/c6dab404-e67c-4849-9ba2-5b1813f1d65c">

DaniGuardiola replied:
Fixed.

DaniGuardiola committed:
Fix downstream issues.

DaniGuardiola committed:
Use ResizeObserver.

DaniGuardiola committed:
Add width transition.

tyxla commented on packages/preferences/src/components/preferences-modal-tabs/style.scss:
```diff
@@ -1,12 +1,15 @@
 
 .preferences__tabs-tablist {
-	position: absolute;
+	position: absolute !important;
```
Would be nice if we could avoid these `!important` rules.

DaniGuardiola replied:
Agreed - unfortunately, emotion is setting its styles with a higher precedence than those set by consumers, so it is hard to override without doing this. I can try using specificity tricks or something like that but I'm not sure they'll work. I wonder if I'm missing something though, as these overrides seem common enough for the team to have realized this style precedence issue before? cc @mirka as she was also surprises during a pairing session the other day.

tyxla replied:
We do have control over that `position: relative` that Emotion sets though, right? I think @mirka suggested using `isolation: isolate` instead of `position: relative` and if I understand correctly, it should fix that issue and also remove the need to use `!important` here.

DaniGuardiola replied:
Yes, that's true for this instance and it's definitely a cleaner / more explicit implementation, so I'll go for that 👍 

The other instance remains though, so I still wonder about the precedence issue. Did you know about it beforehand? Has there been some prior discussion of this I can read?

tyxla replied:
I guess you're talking about `content: none !Important`? 

Won't that also be resolved once we use a shared animation backdrop like we do for `ToggleGroupControl`? Then for the preferences modal tabs, we can just hide the backdrop.

DaniGuardiola replied:
Yes, that's also correct. I still wonder about the precedence issue though.

tyxla replied:
Well, getting rid of Emotion will rid us of those issues 😉

DaniGuardiola replied:
Need to stick with `!important` unfortunately. Not an issue though, since it makes no sense to expect further overrides. Can remove once we migrate off Emotion.

tyxla reviewed (commented):
Is there a good reason why we're not using [shared layout animations](https://www.framer.com/motion/layout-animations/#shared-layout-animations) for the indicator?

We are already using them for `ToggleGroupControl`, see https://github.com/WordPress/gutenberg/pull/50278.

DaniGuardiola commented:
@tyxla that is a nice API! I didn't know about the use of Framer Motion in this package.

I'm hesitant about [adding 40.2kb (gzipped) to the bundle size](https://bundlejs.com/?q=framer-motion&treeshake=%5B%7B+motion+%7D%5D), especially for "nice to have" effects like these. I think the it is simple enough to be implemented with ResizeObserver + CSS in what probably amounts to less than half a kilobyte (that's about 80 times smaller than the `motion` component). I don't know how mindful we need to be of bundle size though.

I'm curious, is there an argument for using it, other than the fact we're already using it somewhere else? I think this is something that could be worth simplifying in the long run, shaving ~40kb off the bundle seems like a huge win to me!

Related thought - the animation that you linked seems very similar to this one, so I wonder if a very small utility could be extracted from my code here and used for both. That'd get the bundle size down even further while keeping things simple.

tyxla commented:
> I'm hesitant about [adding 40.2kb (gzipped) to the bundle size](https://bundlejs.com/?q=framer-motion&treeshake=%5B%7B+motion+%7D%5D), especially for "nice to have" effects like these. I think it is simple enough to be implemented with ResizeObserver + CSS in what probably amounts to less than half a kilobyte (that's about 80 times smaller than the `motion` component). I don't know how mindful we need to be of bundle size though.
> 
> I'm curious, is there an argument for using it, other than the fact we're already using it somewhere else? I think this is something that could be worth simplifying in the long run, shaving ~40kb off the bundle seems like a huge win to me!

Yes, we're mindful of bundle size, however, the `framer-motion` is already in use in this package and for the exact same purpose. The package isn't tree-shakeable right now (this is a different problem to solve), so whether you're using it in one more place or not, it won't make a difference. So my counter-argument is: when we already use a solution and it won't add extra weight but will reduce the amount of custom code, I prefer going with it. One could even argue that adding custom code will add more weight than using what we're already using.

> Related thought - the animation that you linked seems very similar to this one, so I wonder if a very small utility could be extracted from my code here and used for both. That'd get the bundle size down even further while keeping things simple.

Now that we're ending up doing the same kind of shared animation multiple times, abstracting it could indeed be a good idea for a follow-up.

DaniGuardiola commented:
> when we already use a solution and it won't add extra weight but will reduce the amount of custom code, I prefer going with it. One could even argue that adding custom code will add more weight than using what we're already using.

That's fair, though one could then argue that it adds to the technical debt that will make transitioning away from it harder in the future. Right now the argument is "we already use X so might as well use it here", which is totally reasonable, but has the effect that in the future, when the idea of phasing "X" out is proposed, the argument "we use X in many places so it's an expensive refactor we can't afford" will have more weight.

I don't know where the team philosophy sits in regards to this. I've taken a look around all uses of `motion` in this package and they all seem simple(ish) to replace with modern CSS. The popover one, for example, can be replaced with something like [what I did here](https://github.com/Create-Inc/archetype/blob/main/packages/archetype-ui/src/popover/Popover.scoped.css#L15-L21) (the output of this is very small). The workaround above is not necessary anymore as it's been fixed in Ariakit btw. And [I proposed another feature](https://github.com/ariakit/ariakit/issues/2659) that makes it even easier. In any case, what I wonder is whether the team would be on board to formally setting the goal of removing the framer-motion dep - and in that case 1. what's the process of proposing and achieving consensus on something like this? 2. would it then be an okay compromise to add these 0.5kb now to avoid creating technical debt towards the -40kb move?

tyxla commented:
> That's fair, though one could then argue that it adds to the technical debt that will make transitioning away from it harder in the future. Right now the argument is "we already use X so might as well use it here", which is totally reasonable, but has the effect that in the future, when the idea of phasing "X" out is proposed, the argument "we use X in many places so it's an expensive refactor we can't afford" will have more weight.

This is always the case for refactoring, though, so I'll disagree with you, since it's not a strong argument for me. [It is a marathon, not a sprint](https://automattic.com/creed/).

> I don't know where the team philosophy sits in regards to this. I've taken a look around all uses of `motion` in this package and they all seem simple(ish) to replace with modern CSS. The popover one, for example, can be replaced with something like [what I did here](https://github.com/Create-Inc/archetype/blob/main/packages/archetype-ui/src/popover/Popover.scoped.css#L15-L21) (the output of this is very small). The workaround above is not necessary anymore as it's been fixed in Ariakit btw. And [I proposed another feature](https://github.com/ariakit/ariakit/issues/2659) that makes it even easier. In any case, what I wonder is whether the team would be on board to formally setting the goal of removing the framer-motion dep - and in that case 1. what's the process of proposing and achieving consensus on something like this? 2. would it then be an okay compromise to add these 0.5kb now to avoid creating technical debt towards the -40kb move?

I'm happy to remove all dependencies, of course. However, I'd advocate making small, step-by-step improvements. Adding the animation with whatever we're already doing/using feels the most straightforward to me. 

Then, as a next step, if we want to refactor away from `framer-motion`, we could experiment with that, but note that the `components` package [re-exports `motion`](https://github.com/WordPress/gutenberg/blob/10859af537871d0010f7e904705a1672aad9cc78/packages/components/src/animation/index.tsx#L11) (albeit as unstable API), and it's utilized broadly across the editor. So refactoring away from `framer-motion` completely will most likely require thorough testing in a separate effort. 

Let me know what you think.

DaniGuardiola committed:
Simplify and use framer motion

DaniGuardiola committed:
vertical indicator

DaniGuardiola commented:
Update: I refactored this to use the Framer Motion lib, and unfortunately there's a ton of problems when testing the instances in a wp instance (you can load this PR in the playground to see them). I've spent quite a bit of time today trying to find a decent solution but nothing's working, so I might take a small break from this PR, especially since I'm working on adding support for loading specific Gutenberg commits to the playground - a feature that will be very helpful in debugging this.

tyxla commented:
> Update: I refactored this to use the Framer Motion lib, and unfortunately there are a ton of problems when testing the component instances in a wp instance (you can load this PR in the playground to see them). I've spent quite a bit of time today trying to find a decent solution but nothing's working, so I might take a small break from this PR, especially since I'm working on adding support for loading specific Gutenberg commits to the playground - a feature that will be very helpful in debugging this.

I'd love to learn more about the problems you encountered. I gave this some testing in a bunch of places, but the only weird behavior I stumbled upon were: 

1. With vertically oriented tabs, and in this case, I think we just don't need the indicator at all:

https://github.com/WordPress/gutenberg/assets/8436925/ee8dd7af-ac1f-4855-a06e-5a0c0274893d

2. With horizontally oriented tabs in some contexts, like when setting text background color, which most likely happens due to the fact that the different settings for text color, background color and link color share the same `layoutId`:

https://github.com/WordPress/gutenberg/assets/8436925/da9d0c8c-fd89-4859-8a97-02f1404142ed

Both of these seem fixable, although I haven't had a chance to play with them. Are there any other problems you encountered?

DaniGuardiola commented:
Thanks for looking into it @tyxla!

1 is just a TODO, but 2 is indeed one of the issues. For that one, I've tried to debug it in multiple ways, but always assumed that the `layoutId` was unique to the instance since I'm using the `instanceId` from the context, which seemed adequate after reading its docs: "The unique id string for this instance of the Tabs component". Am I missing something? Can these be duplicated between Tabs instances?

The other issue I've noticed is that there's an annoying animation on page load on some instances of Tabs, likely because there is some sort of immediate re-render that Framer Motion is picking up and animating.

You can observe it here, I'm constantly reloading the page in this recording:

https://github.com/WordPress/gutenberg/assets/8862967/aff733ea-11bd-44a5-b261-8423619b16a2

I also spent a good amount of time trying to figure out a "clean" universal solution that isn't a dirty hack, but without control of the animation it's a bit hard to come up with something other than a hacky initial `setTimeout`-based delay. I think there is a good way to fix it when controlling the animation, since it should be possible to only enable transitions when the selected tab changes, and making the indicator resize instantly when the tab does instead.

If you have any ideas, I'd be happy to hear them, as I feel a bit stuck!

tyxla commented:
> For that one, I've tried to debug it in multiple ways, but always assumed that the layoutId was unique to the instance since I'm using the instanceId from the context, which seemed adequate after reading its docs: "The unique id string for this instance of the Tabs component". Am I missing something? Can these be duplicated between Tabs instances?

That's a correct assumption, since that's how the tabs context is implemented, but it's worth verifying case-by-case that it's actually true. Maybe we're inadvertently reusing the same instances in multiple places. 

An alternative in the case of duplication is to use wrapping `LayoutGroup` that will serve as namespaces for the shared animations.

> The other issue I've noticed is that there's an annoying animation on page load on some instances of Tabs, likely because there is some sort of immediate re-render that Framer Motion is picking up and animating.

Oh, that one. That occurs because during initial load, the first tab name is called "Document":

https://github.com/WordPress/gutenberg/blob/486b2dba17213e5117f4a6495160f443c9b10184/packages/edit-post/src/components/sidebar/settings-header/index.js#L24

And once the current post type data loads, it changes the label from "Document" to the name of the post type ("Post" in your example). 

Because "Document" is longer than "Post", you'll see `framer-motion` animating from a longer border to a smaller one. That's actually quite accurate behavior, if you ask me. We might want to disable that animation somehow, although I'll have to tinker with it to actually suggest a proper solution (I likely won't have time for that today). Let me know if this hint is useful though.

DaniGuardiola commented:
> it's worth verifying case-by-case that it's actually true. Maybe we're inadvertently reusing the same instances in multiple places.

I verified that this is not the case, instance IDs for all rendered tabs when reproducing these issues are unique. What do you mean by "reusing the same instances"? I haven't looked at the source, but my understanding is that these IDs are unique to the React **element**, is this correct?

> That occurs because during initial load, the first tab name is called "Document"

Yes, I understand why this happens. While it is "correct", I don't think it's what we want, I believe what we want is to animate the indicator when transitioning between different tabs. I've been looking at Framer Motion and haven't been able to find any granular way to control (i.e. enable/disable) the animation. I know how I could do it on my previous implementation though, and it is pretty straightforward, so it's a shame that Framer Motion doesn't support this. Maybe I just missed it, I was actually surprised when I couldn't find it.

---

By the way, I'm not sure why because the context and code is exactly as it was when I last worked on this, but now I'm seeing a new buggy behavior:


https://github.com/WordPress/gutenberg/assets/8862967/7987b019-3e49-464e-9f6f-e5d19b024ea2

It doesn't jump from the opposite direction like previously, but now it jumps in a glitchy, incorrect way.

This one's a real head-scratcher :(

DaniGuardiola committed:
Revert to previous implementation.

DaniGuardiola committed:
Fix bug due to some animations breaking measurement of the tab element.

DaniGuardiola committed:
Abstracted and fixed all previous issues.

DaniGuardiola commented:
@tyxla @mirka FYI:

- Reverted to previous approach.
- Abstracted into lightweight utilities for clarity and potential re-use.
- Fixed all known issues (all glitchy behavior showcased in previous messages, including unwanted transitions) plus a couple more I found.
- Had to revert back to the `!important` strategy for overrides, unfortunately `isolation: isolate` doesn't work for this purpose either.
- Gotta make it work for vertical orientation - half the code is there though.
- Will do another testing pass tomorrow to make sure everything's in order.
- I will also ensure that all comments have been addressed. For example, @mirka discussed focus-related animations.
- No need to review yet if you don't want to since there are some pending tasks.

Semi-related: https://github.com/WordPress/gutenberg/issues/60975

jsnajdr commented on packages/components/src/tabs/styles.ts:
```diff
@@ -13,12 +13,35 @@ import { space } from '../utils/space';
 	&[aria-orientation='vertical'] {
 		flex-direction: column;
 	}
+	&.isAnimationEnabled::after {
```
nit: the CSS class naming convention used across Gutenberg is more like `.is-animation-enabled`.

mirka replied:
More info about our BEM dialect in the [contributing guidelines](https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/coding-guidelines.md).

DaniGuardiola replied:
This is a conscious decision I made for authoring components a long time ago, after getting annoyed with this kind of thing:

```ts
<div className={clsx({
  "is-animation-enabled": isAnimationEnabled,
  "is-whatever": isWhatever,
  "is-this": isThis,
  "is-that": isThat,
})} />
```

Doing camel case unlocks this pattern:

```ts
<div className={clsx({
  isAnimationEnabled,
  isWhatever,
  isThis,
  isThat,
})} />
```

That said, I agree I should be following the convention. I might propose this as a broader change in the future though.

tyxla replied:
The CSS class naming conventions are inherited from (and compatible with) the [WordPress CSS coding standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/#selectors). Changing those at this time is a near-impossible task.

jsnajdr commented:
> like the site editor sidebar animation work @jsnajdr did recently in #60849

One thing I still strongly dislike about migrating that animation from `motion.div` to CSS animation: the keyframes name is global and pollutes the global namespace. In theory we should be able to do this:
```css
@keyframes :local(slide-from-left) {
  ...
}

.has-slide-from-left-anim {
  animation-name: slide-from-left;
}
```
The webpack `css-loader` should be able to interpret the `:local` syntax and mangle the names. But it currently doesn't work. Most likely something in the SCSS/postcss/css-loader build pipeline needs to be properly configured to make it work.

mirka commented:
> The webpack `css-loader` should be able to interpret the `:local` syntax and mangle the names. But it currently doesn't work. Most likely something in the SCSS/postcss/css-loader build pipeline needs to be properly configured to make it work.

Last time I checked, stylesheets don't go through webpack, but a custom build script instead. So a lot of features that are available in css-loader out of the box can't be used until we address that.

tyxla commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -17,19 +23,106 @@ import type { TabListProps } from './types';
+	previousValue: unknown;
+};
+
+function useOnValueUpdate(
+	value: unknown,
+	onUpdate: ( context: ValueUpdateContext ) => void
+) {
+	const [ previousValue, setPreviousValue ] = useState( value );
+
+	useEffect( () => {
+		if ( previousValue !== value ) {
+			onUpdate( { previousValue } );
+			setPreviousValue( value );
+		}
+	}, [ previousValue, value, onUpdate ] );
+}
```
Why is this abstraction necessary? Can't we just rely on `selectedId` being a dependency of `useEffect` and when it changes, it'd retrigger the effect? Let me know if I'm missing something

DaniGuardiola replied:
It needs to only trigger on updates (not initially), and knowledge of the previous value is necessary to bail on some updates too. The useEffect dep array is not enough. This is a common hook util I've needed sooner or later on every design system I've built :P

tyxla replied:
> It needs to only trigger on updates (not initially)

That can be done with a simple ref, can't it?

>  and knowledge of the previous value is necessary to bail on some updates too

Can you speak more about that? Am I missing the use case for this in the code?

DaniGuardiola replied:
This abstraction is used later on in the component to enable transitions only when the selected tab id is updated, which stops unwanted transitions from happening. It is not enough to check for the element to be updated.

For example, remember the Document -> Post re-render that was causing an unwanted animation. In that re-render, the element is swapped, yet the tab id is stable. Hence a pointer to an element (which is what I assume you mean by "ref", though it cannot be a React ref in this case since we want to schedule actual state updates here) is not enough.

Finally, what I meant by the "knowledge of the previous value" bit is that, in some cases, some Tabs instances (e.g. the color picker ones) initially start with a "null" id as selected, which is then replaced by an actual id when a re-render happens. This used to trigger a transition from "nothing" to the initially selected tab. This is avoided by bailing out of enabling the transition when the previous value was "null" (or just falsy in general - which is okay because we just want to ensure it was an id/string). You can see this check here: https://github.com/WordPress/gutenberg/pull/60560/files/283f10952c38cf35c184042ed9e11134e2fd1d71#diff-64263a335082a1c717edd91225f49242f93a3b2625646af8af012dc96a6ff7a6R116 (`previousValue &&`)

jsnajdr replied:
The hook is fine, but you'd be better off storing the previous value in a ref:
```js
const previousValue = useRef( value );
useEffect( () => {
  if ( previousValue.current !== value ) {
    onUpdate( { previousValue: previousValue.current } );
    previousValue.current = value;
  }
}, [ value, onUpdate ] );
```
`useState` triggers an unnecessary extra render when `value` changes, but `useRef` doesn't.

The hook could be also more precisely typed:
```tsx
useOnValueUpdate<T>(value: T, onUpdate: (context: ValueUpdateContext<T>) => void)
```

One day, if used widely, it could be moved to `@wordpress/compose`. There is a `usePrevious` hook there which is somewhat related.

Another slight improvement is storing the `onUpdate` callback in a ref, so that the value-updating effect is triggered only by `value` change and nothing else. Note that the effect is a noop anyway when `onUpdate` changes.
```js
const latestOnUpdate = useRef( onUpdate );
useEffect( () => {
  latestOnUpdate.current = onUpdate;
}, [ onUpdate ] );

const previousValue = useRef( value );
useEffect( () => {
  if ( previousValue.current !== value ) {
    latestOnUpdate.current( { previousValue: previousValue.current } );
    previousValue.current = value;
  }
}, [ value ] );
```

DaniGuardiola replied:
Thank you for the notes @jsnajdr. I already have these changes locally (not yet pushed as I'm debugging something), you read my mind there :P

While the function ref works fine like that, I tend to use a layout effect just to make sure it runs as early as possible. It's what is [typically used in `useEventCallback` implementations](https://github.com/facebook/react/issues/14099#issuecomment-440013892) too (and you can also often see it without abstractions, e.g. [see Kent C. Dodds take](https://epicreact.dev/the-latest-ref-pattern-in-react/)). Hopefully we'll just be able to use [something like `useEvent`](https://github.com/reactjs/rfcs/pull/220) if it officially lands in React.

DaniGuardiola replied:
Btw, I'm already doing the function ref thing a few lines above: https://github.com/WordPress/gutenberg/pull/60560/files/283f10952c38cf35c184042ed9e11134e2fd1d71#diff-64263a335082a1c717edd91225f49242f93a3b2625646af8af012dc96a6ff7a6R40

jsnajdr replied:
> Btw, I'm already doing the function ref thing a few lines above:

I think it's enough when the `useOnValueUpdate` hook does it. In the calling component, IMO it's a bit pointless.

DaniGuardiola replied:
It's not used in the calling component though :P

jsnajdr replied:
Oh, disregard my previous comment, I totally misread the code. I thought that it stores the `onUpdate` callback to a ref and then calls `useOnUpdateValue` with the ref's `.current` value.

DaniGuardiola replied:
No worries!

jsnajdr commented:
> Last time I checked, stylesheets don't go through webpack, but a custom build script instead.

Then there is this [postcss-modules](https://github.com/madyankin/postcss-modules) package that processes the `:local` and `:global` syntax and does the mangling. We should try it out 🙂 

It'll be even more useful if we start abandoning Emotion, which produces unique mangled CSS class names.

tyxla commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -17,19 +23,106 @@ import type { TabListProps } from './types';
+		}
+
+		// Set up a ResizeObserver.
+		if ( ! resizeObserverRef.current ) {
```
I wonder if it makes sense to use our `useResizeObserver()` hook?

DaniGuardiola replied:
It has a different API that doesn't work here.

tyxla replied:
Can you please elaborate on why it doesn't work here?

DaniGuardiola replied:
That hook does a lot of things that I'm sure are useful for a use-case that I'm not familiar with (tracking the size of an element is a nuanced endeavour). Without documentation, it's really hard to know exactly what this use case is, and what the options do, but I found a few critical things that make it incompatible with what I need:

- It only works with refs. Not sure exactly how, as it seems to return a callback ref, but the docs indicate that you should put it as a child of a React element. Not sure how that's supposed to work, but in any case I need to track multiple elements so the element input can't be a ref because it needs to be statefully tracked to change the observed target.
- It seems to return only the size, while I need positioning information.
- The size it returns is not the size I need. My hook is resilient to transformation effects, for example, which makes a difference in tabs contained in popups during entry/exit transitions. What I need is the "offset" position and size, because the indicator is relative to the tab container, locally.

This hook uses a ResizeObserver because that's the appropriate platform API for the use case, but it does something very specific: track the position and size of the offset rect of a given element (that can change). That's why it's called `useTrackElementOffset`. Thinking of renaming it to `useTrackElementOffsetRect` for enhanced precision.

```ts
function useTrackElementOffset(
	targetElement?: HTMLElement | null,
	onUpdate?: () => void
): {
    left: number;
    top: number;
    width: number;
    height: number;
}
```

Takes a target element in, returns stateful and always-up-to-date position and size of the offset rect. This is precisely what I need for this use case. The update callback argument is optional, and I don't even use it here but it seems like a good idea if this is used as a more general utility, though I don't want to abstract too early so for now I'm keeping it here, local to `Tabs`.

tyxla commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -17,19 +23,106 @@ import type { TabListProps } from './types';
+
+		// Unobserve previous element.
+		const { current: observedElement } = observedElementRef;
+		if ( observedElement ) resizeObserver.unobserve( observedElement );
```
I was going to suggest some cleanup on unmount. However, if we end up using `useResizeObserver()`, it does its own cleanup already.

DaniGuardiola replied:
Will add a cleanup.

DaniGuardiola committed:
Follow naming convention for classes.

DaniGuardiola committed:
Support vertical orientation + misc fixes and improvements.

DaniGuardiola committed:
Clean up styles a bit.

DaniGuardiola committed:
Better focus ring animation + minor style cleanup.

DaniGuardiola committed:
Merge branch 'trunk' into tabs-animation

DaniGuardiola committed:
Fix changelog (oops).

DaniGuardiola committed:
Actually fix changelog.

DaniGuardiola commented:
fyi ready for a final review @mirka @tyxla @jsnajdr

stokesman commented on packages/preferences/src/components/preferences-modal-tabs/style.scss:
```diff
@@ -1,12 +1,15 @@
 $vertical-tabs-width: 160px;
 
 .preferences__tabs-tablist {
```
```suggestion
[class].preferences__tabs-tablist {
```
Is a strong enough specificity hack to obviate `!important` usage here but this is pretty nit-picky given there’s likely no one trying to override those rules. Mostly sharing because it’s a handy hack in general.

DaniGuardiola replied:
Yeah, in fact `.preferences__tabs-tablist.preferences__tabs-tablist` works. However, I chose this explicitly for two reasons: 1. it is more obvious that we're forcing an override and 2. the root issue is that the component styles should not have precedence in the first place, so this code's readability should not suffer from it. As a bonus, doing it that way would be fragile since it could stop working if the underlying styles (which are an implementation details) starts using a selector with higher specificity for some reason.

Thanks for chiming in :)

DaniGuardiola committed:
Remove deprecated `reduceMotion` utility.

DaniGuardiola commented:
Context for [last commit](https://github.com/WordPress/gutenberg/pull/60560#:~:text=2-,Remove%20deprecated%20reduceMotion%20utility.,-6b2f541): https://github.com/WordPress/gutenberg/issues/60902#issuecomment-2078058266

DaniGuardiola commented:
Gently pinging @WordPress/gutenberg-components for a review :)

tyxla commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -17,19 +23,115 @@ import type { TabListProps } from './types';
 import { TabListWrapper } from './styles';
 import type { WordPressComponentProps } from '../context';
+import type { CSSProperties } from 'react';
+import classNames from 'classnames';
```
This needs a refresh after we switched to `clsx`.

tyxla commented on packages/components/src/tabs/styles.ts:
```diff
@@ -10,15 +10,46 @@ import * as Ariakit from '@ariakit/react';
+		border-bottom: var( --wp-admin-border-width-focus ) solid
+			${ COLORS.theme.accent };
+	}
+	&[aria-orientation='vertical']::after {
```
Unrelated to the PR: I think we could use a story with vertical tabs. It's a very valid use case that's not represented in Storybook. Perhaps a good idea for a follow-up?

DaniGuardiola replied:
Added, it'll be useful as I tweak it.

tyxla commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -50,9 +152,28 @@ export const TabList = forwardRef<
+				{
+					'--indicator-left': `${ indicatorPosition.left }px`,
+					'--indicator-top': `${ indicatorPosition.top }px`,
+					'--indicator-width': `${ indicatorPosition.width }px`,
```
Spotted this indicator width/spacing weirdness, which seems to be pre-existing:

![Screenshot 2024-05-08 at 17 48 37](https://github.com/WordPress/gutenberg/assets/8436925/018d3077-fd9b-417a-836b-2973c6c90e80)
![Screenshot 2024-05-08 at 17 48 44](https://github.com/WordPress/gutenberg/assets/8436925/dd1b6820-84e8-424f-a66f-529dbae0ca56)

See how there is some padding on the right of the Gradient item. Definitely something we should try fixing.

You can see that interface by editing global styles, and navigating to Colors > Button.

DaniGuardiola replied:
I think that's just a consequence of the min-width of the container being larger than the natural width of the tab list. You can see this example where it also happens:

<img width="564" alt="image" src="https://github.com/WordPress/gutenberg/assets/8862967/1915c625-d7b8-49ab-bcf4-9bdef0a51a1f">

The only difference is that in your example, by pure luck, the width is _almost_ the same as the container, so I see how it might feel like there's some weird padding, but it's simply a natural result of sizing.

tyxla commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -50,9 +152,28 @@ export const TabList = forwardRef<
+					'--indicator-top': `${ indicatorPosition.top }px`,
+					'--indicator-width': `${ indicatorPosition.width }px`,
+					'--indicator-height': `${ indicatorPosition.height }px`,
+				} as CSSProperties
```
Not a blocker, but still, can we figure out a better way to handle CSS var types in the future?

DaniGuardiola replied:
This is relevant: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/8871b1e8938a1ed698ec8a88c77ee169294e45d4/types/react/index.d.ts#L974-L983

Which links to https://github.com/frenic/csstype#what-should-i-do-when-i-get-type-errors

If we don't want to constrain CSS custom properties, we could go with a simple augmentation:

```ts
// My css.d.ts file
import type * as CSS from 'csstype';

declare module 'csstype' {
  interface Properties {
    // Allow any CSS Custom Properties
    [index: `--${string}`]: any;
  }
}
```

I think this is the way to go for a proper fix. For one-offs, that README suggests casting the property itself:

```ts
const style: CSS.Properties = {
  // Add a CSS Custom Property
  ['--theme-color' as any]: 'black',
};
```

While this is technically more type-safe, it's also both more cumbersome and will trigger eslint which will need to be silenced. So I usually just reach for casting. I think I'll create a PR with the global fix (augmentation approach with `--` prefix) to gather feedback.

DaniGuardiola replied:
A solution has been merged, so I'll update this PR shortly: https://github.com/WordPress/gutenberg/pull/61872

tyxla commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -17,19 +23,115 @@ import type { TabListProps } from './types';
+		const { current: observedElement } = observedElementRef;
+		if ( observedElement ) resizeObserver.unobserve( observedElement );
+
+		// Observe new element.
```
Instead of unobserving every time and re-observing if the element is still there, should we just do it at proper unmount cleanup (the returned callback of `useEffect()`)?

DaniGuardiola replied:
Using an effect cleanup might actually be better, I think it is currently leaking on unmount. Let me try, and remember if there's a reason why I didn't do it like that.

DaniGuardiola replied:
Seems to work properly!

tyxla commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -17,19 +23,115 @@ import type { TabListProps } from './types';
 	const context = useTabsContext();
+
+	const selectedId = context?.store.useState( 'selectedId' );
+	const indicatorPosition = useTrackElementOffset(
+		context?.store.item( selectedId )?.element
+	);
+
+	const [ animationEnabled, setAnimationEnabled ] = useState( false );
+	useOnValueUpdate(
+		selectedId,
+		( { previousValue } ) => previousValue && setAnimationEnabled( true )
+	);
```
Much/all of the animation/indicator magic seems to be unnecessary if we're hiding the indicator. Should the animated indicator be optional then?

DaniGuardiola replied:
I don't see what is gained by adding an option. The impact of the animation is minimal, especially if hidden through a CSS override. Or am I missing your point?

tyxla replied:
I agree, it's pretty harmless to keep as-is. No need for an option.

tyxla commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -17,19 +23,115 @@ import type { TabListProps } from './types';
+		context?.store.item( selectedId )?.element
+	);
+
+	const [ animationEnabled, setAnimationEnabled ] = useState( false );
```
Looking at the code, I'd always expect animations. However, there are cases when animations won't be there, for example:

- With reduced motion
- For vertically oriented tabs

Should we document these and set the right expectations that animations are expected only for vertical tabs and when `prefers-reduced-motion` is not enabled?

DaniGuardiola replied:
No animations with `prefers-reduced-motion` should be expected in every place where we have animations, and it's made explicit in the CSS. I don't personally feel like any further documentation is necessary in that front, unless we decide to do it consistently. However, we don't really have user-facing documentation in the first place for individual components anyway :P

Vertical tabs actually do have an animation. I might change this if we don't want the change (as discussed somewhere else in this review's comment), but currently it does, and it works in the same way except vertically (in the right side).

tyxla commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -17,19 +23,115 @@ import type { TabListProps } from './types';
+	previousValue: T;
+};
+
+function useOnValueUpdate< T >(
+	value: T,
+	onUpdate: ( context: ValueUpdateContext< T > ) => void
+) {
+	const previousValueRef = useRef( value );
+
+	// TODO: replace with useEventCallback or similar when officially available.
+	const updateCallbackRef = useRef( onUpdate );
+	useLayoutEffect( () => {
+		updateCallbackRef.current = onUpdate;
+	} );
+
+	useEffect( () => {
+		if ( previousValueRef.current !== value ) {
+			updateCallbackRef.current( {
+				previousValue: previousValueRef.current,
+			} );
+			previousValueRef.current = value;
+		}
+	}, [ value ] );
+}
```
Why do we need to maintain the update callback in a ref? What difference would it make if I did:

```suggestion
function useOnValueUpdate< T >(
	value: T,
	onUpdate: ( context: ValueUpdateContext< T > ) => void
) {
	const previousValueRef = useRef( value );

	useLayoutEffect( () => {
		if ( previousValueRef.current !== value ) {
			onUpdate( {
				previousValue: previousValueRef.current,
			} );
			previousValueRef.current = value;
		}
	}, [ value, onUpdate ] );
}
```

DaniGuardiola replied:
This is a low-level event-oriented hook that needs to fire only on value updates. If the callback was a dependency for the effect, passing a different callback would trigger it, ergo the hook would not be firing on value updates exclusively. That would be misleading, and would also break the logic of consumers. In our case, such scenario would cause the animation to be enabled at a time when it shouldn't be.

DaniGuardiola replied:
Btw, @jsnajdr also proposed this as I was about to push a commit with it, we had a parallel thinking moment :P https://github.com/WordPress/gutenberg/pull/60560#discussion_r1577505749

tyxla commented on packages/components/src/tabs/tablist.tsx:
```diff
@@ -17,19 +23,115 @@ import type { TabListProps } from './types';
+				width: element.offsetWidth,
+				height: element.offsetHeight,
+			} );
+			updateCallbackRef.current?.();
```
What difference does it make if I just call `onUpdate?.()` here (and add it as a dependency of course)? It's not immediately clear why we need the `updateCallbackRef` in the first place.

DaniGuardiola replied:
Similar logic to https://github.com/WordPress/gutenberg/pull/60560#discussion_r1610112412, though in this case due to how the effect is implemented (idempotently) it wouldn't really break anything, but it would cause wasteful computation since there's an unnecessary dependency for the effect.

tyxla commented on packages/preferences/src/components/preferences-modal-tabs/style.scss:
```diff
@@ -1,12 +1,15 @@
 	left: $grid-unit-20;
 	width: $vertical-tabs-width;
 
+	&::after {
```
Is this something we need to handle in the component itself? Perhaps if we provide an API to declare the indicator to be optional?

DaniGuardiola replied:
This seems like an escape hatch and it's only used here. You could also say the same about everything else that's being customized or overriden here (a lot). I would advise against supporting this to prevent configuration-creep from building up. If this pattern becomes "official" or, at least, common enough, then I would agree.

At the root of this I think the problem is that `Tabs` from `@wordpress/components` is not what was needed here. Rather, it seems like we need tabs that look a certain way, and for that it seems like using Ariakit with the styles wanted for this interface is the way to go.

tyxla reviewed (commented):
Thanks for the great work here! 

This actually appears to work really well in my testing in the editor. 

I wish there was a vertical storybook instance for tabs. It's easy enough to toggle the `orientation` prop, but it's also easy to miss a vertical tab detail when working on `Tabs`.

With the vertical tabs in mind, this does introduce some changes which I'm not sure are intentional. If you take a look at the "With Tab Icons And Tooltips" story, you'll see the indicator at the bottom in `trunk`:

![Screenshot 2024-05-08 at 18 47 09](https://github.com/WordPress/gutenberg/assets/8436925/08ad6b28-f2a2-49d7-8b66-70fb90394c75)

but in this PR, the indicator is on the right:

![Screenshot 2024-05-08 at 18 47 16](https://github.com/WordPress/gutenberg/assets/8436925/5644b1d4-2259-43d8-bd36-7d1c03bae542)

and I'm not sure that's intentional. 

I think CSS animations are the way to go here, but I'm curious if we can simplify some parts of the code, so I've asked some naive questions here and there and made some suggestions. 

Let me know what you think!

DaniGuardiola committed:
Merge branch 'trunk' into tabs-animation

DaniGuardiola committed:
Fix open/closed

DaniGuardiola committed:
Add vertical tabs story

DaniGuardiola committed:
Move ResizeObserver unobserve to effect cleanup

DaniGuardiola commented:
@tyxla check the new vertical tabs story to see the current look. I think there's a point to reverting it to how it worked before (horizontal and no animation), but take a look first just so we're on the same page :)

tyxla commented:
I've done some extensive testing in the post editor, site editor and storybook and things look great, aside from the RTL support (see below).

Code also looks good - thanks for taking the time to explain some of the decisions!

> @tyxla check the new vertical tabs story to see the current look. I think there's a point to reverting it to how it worked before (horizontal and no animation), but take a look first just so we're on the same page :)

I think it works great, thanks! 

The only thing I can see not working well is RTL support - right now the indicator will appear on the right for RTL, which is undesired:

LTR:
![Screenshot 2024-05-23 at 10 16 08](https://github.com/WordPress/gutenberg/assets/8436925/4bc8ed74-0afd-41fc-aca0-ab76d27d1cfd)

RTL:

![Screenshot 2024-05-23 at 10 16 22](https://github.com/WordPress/gutenberg/assets/8436925/f97139ac-360d-4070-8240-22f3164c8d27)

Also, can we get a pair of eyes from @WordPress/gutenberg-design before we ship it?

jasmussen commented:
Nice work. The horizontal animation where the bar moves to the tab that gets selected, that works great, as it echoes the segmented control backdrop animation:

![segmented](https://github.com/WordPress/gutenberg/assets/1204802/3a79e644-0bb2-4ec6-961b-e3ad9ceaf5f3)

It should probably also use the same animation metrics, speed, bounce, etc — the idea being, these are the same physics, just a different visual.

I think @youknowriad explored an alternative to framer motion recently, would be good to double check first.

I would not add the _line_ to these vertical tabs:

![Screenshot 2024-05-23 at 09 46 53](https://github.com/WordPress/gutenberg/assets/1204802/6abf18b6-a39d-4093-a5e5-e7a5650f45fe)

If we add animation to the vertical tabs, it should be the gray backdrop itself that moves, just like the backdrop on the segmented control moves.

DaniGuardiola commented:
@jasmussen thanks for chiming in!

> It should probably also use the same animation metrics, speed, bounce, etc — the idea being, these are the same physics, just a different visual.

Agreed. Since I'm gonna be re-using some of the logic to rework some of those components as a follow-up, I will make sure then that it uses the same animation parameters 👍 

Re: Framer Motion, this uses a vanilla CSS transition, with only some minimal logic to track position and size of the indicator and some tweaks to prevent unintended animations. The goal is to achieve the same, potentially with some code re-use, in those other components as well. Related: #60975 

> If we add animation to the vertical tabs, it should be the gray backdrop itself that moves, just like the backdrop on the segmented control moves.

I think that's a great idea. No RTL concerns either with that approach cc @tyxla. I'm gonna experiment with it and ping back for feedback.

tyxla commented:
> I think that's a great idea. No RTL concerns either with that approach cc @tyxla. I'm gonna experiment with it and ping back for feedback.

Sounds good to me as well. One way to approach it is to leave vertical tabs without animation, and address it in a separate PR, to keep changes as contained as possible.

tyxla commented:
In terms of the animation itself, I think we can play with a different easing, maybe add a small bit of bounce (a variation of `easeOutBounce` with a reduced bounce), and that would improve the feeling. 

The goal is that the tabs backdrop will use the same CSS animation instead of framer motion, FWIW. Framer animations can be beautiful, but are very heavy on resources and come with too many trade-offs, thus our goal to minimize their use and avoid using them for the components package. Ideally the [View Transitions API](https://developer.mozilla.org/en-US/docs/Web/API/View_Transitions_API) will get a broader support and we'll be able to completely remove the package in favor of those rich, slick, optimized transitions.

DaniGuardiola commented:
@tyxla feel free to suggest a specific easing curve, something like [this](https://cubic-bezier.com/) might help, and you can also use the Chrome DevTools to see it in action directly in the storybook:

https://github.com/WordPress/gutenberg/assets/8862967/8313a9bd-b097-4f88-8cc8-0b7326361c84

I still need to push a couple of remaining tweaks here anyway and it's a one line change, so there's room for one last feedback round :)

tyxla commented:
> @tyxla feel free to suggest a specific easing curve

I'd keep it something simple (without an obvious bounce), like https://easings.net/#easeInOutCubic or https://easings.net/#easeInOutExpo but I have no strong preferences. I'm definitely not an easing expert, so feel free to experiment with different ones. Again, this is something we can do subsequently in another PR.

DaniGuardiola commented:
@tyxla sure! I'm personally a fan of `ease-out` (which I chose for this PR) for "active" user interactions because it provides instant feedback (as opposed to ease-in) but still gives you some easing to make it look smoother and less jarring than a linear curve. I think it's often the best compromise between good feedback to interaction (important for accessibility and good user experience) and "delightfulness". For other components, like buttons, I tend to favor no transitions at all.

Would love your thoughts here too @jasmussen.

None of this is blocking for this PR btw, just want to get the convo going.

jasmussen commented:
My instinct also veers towards ease-out, though a CC to @jameskoster as well, as he's also been diving into this. My main recommendation for the animations in general is that they should feel fluid and snappy. _Fast_. Almost never longer than 0.2 seconds, sometimes 0.1s in speed. The best way to feel this out, is in the branch itself and so far the speed seems fine — and if nothing else, the speed from the segmented control is the model to follow.

I'm also assuming you are already on top of this (and I think it's built-in to the componentry), but will state it to just be sure, all of these animations should respect a users `reduce-motion` setting.

jameskoster commented:
It would probably make sense to tokenise animation effects at some point, we're a bit all over the place at the moment. `easeOut` seems to be common, but there are also have several custom timing functions, notably: 

* Interface skeleton: https://github.com/WordPress/gutenberg/blob/4da057ecd6958d40b8f2a83db1152320e5f161e6/packages/edit-site/src/components/editor/index.js#L320
* DropdownMenu V2: https://github.com/WordPress/gutenberg/blob/f49e0b5958adad9fbd211d5ded07b4cabf5bfec3/packages/components/src/dropdown-menu-v2/styles.ts#L21

For now I agree it makes sense to mimic the segmented control 1:1, then revisit holistically.

DaniGuardiola committed:
Merge branch 'trunk' of https://github.com/WordPress/gutenberg into tabs-animation

DaniGuardiola committed:
Remove outdated type cast.

DaniGuardiola committed:
Hide vertical indicator for now.

DaniGuardiola commented:
Thanks everyone. I've pushed a couple of last details and this should be ready to merge. Feel free to do a final review @tyxla, looking at my latest commits.

tyxla reviewed (approved):
Looks good and tests well! 🚀 

Thanks @DaniGuardiola! 🙌 

Let's follow-up on adding the animation to the vertical tabs when you get a chance.

richtabor commented:
Noticed this line is no longer lined up properly: 

![CleanShot 2024-05-23 at 17 21 45](https://github.com/WordPress/gutenberg/assets/1813435/0b3d22cd-6f90-42cc-95c9-9dfa6fafa0a0)

DaniGuardiola commented:
@richtabor I will take a look at it, thanks for reporting.

ciampo commented:
👋 Catching up with this conversation, I'd be interested in debugging a bit more the issues encountered while using framer motion, as I debugged a few similar ones while working on `ToggleGroupControl`. Were the issues present in isolation (ie Storybook) too, or only when rendering the component in the editor?

One of the main reasons we opted for using `framer-motion` was that the previous custom implementation was buggy and difficult to maintain (especially for edge cases like the component being rendered inside a popover, etc).

Also, since framer motion [is exported from the components package](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/animation/index.tsx), I'm not sure it's going to be possible to remove it entirely as a dependency.

In general, I agree that both for visual consistency and for long-term maintainability, we should aim at having a unified way of implementing such animations across our components.

DaniGuardiola commented:
@jasmussen I added the vertical indicator animation: https://github.com/WordPress/gutenberg/pull/62879
